code/modules/admin/holder2.dm

```diff
@@ -148,6 +173,149 @@ GLOBAL_PROTECT(href_token)
 			return TRUE //we have all the rights they have and more
 	return FALSE
 
+// TRUE for a vaild connection, null is the id (it is unnecessary)
+#define VALID_2FA_CONNECTION list(TRUE, null)
+
+/// Returns whether or not the given client has a verified 2FA connection.
+/// The output is in the form of a list with the first index being whether or not the
+/// check was successful, the 2nd is the ID of the associated database entry
+/// if its a false result and if one can be found.
+/datum/admins/proc/check_2fa(client/client)
+	if (bypass_2fa)
+		return VALID_2FA_CONNECTION
+
+	var/admin_2fa_url = CONFIG_GET(string/admin_2fa_url)
+
+	// 2FA not being enabled == everyone passes
+	if (isnull(admin_2fa_url) || admin_2fa_url == "")
+		return VALID_2FA_CONNECTION
+
+	// I believe this is only in the case of Dream Seeker.
+	if (isnull(client?.address))
+		return VALID_2FA_CONNECTION
+
+	if (!SSdbcore.Connect())
+		if (verify_backup_data(client))
+			return VALID_2FA_CONNECTION
+		else
+			return list(FALSE, null)
+
+	var/datum/db_query/query = SSdbcore.NewQuery({"
+		SELECT id, verification_time FROM [format_table_name("admin_connections")]
+		WHERE ckey = :ckey
+		AND ip = INET_ATON(:ip)
+		AND cid = :cid
+	"}, list(
+		"ckey" = client.ckey,
+		"ip" = client.address,
+		"cid" = client.computer_id,
+	))
+
+	if (!query.Execute())
+		qdel(query)
+		return list(FALSE, null)
+
+	var/is_valid = FALSE
+	var/id = null
+
+	if (query.NextRow())
+		id = query.item[1]
+		is_valid = !isnull(query.item[2])
+
+	qdel(query)
+	return list(is_valid, id)
+
+#undef VALID_2FA_CONNECTION
+
+/datum/admins/proc/start_2fa_process(client/client, id)
+	add_verb(client, /client/proc/admin_2fa_verify)
+	client?.init_verbs()
+
+	var/admin_2fa_url = CONFIG_GET(string/admin_2fa_url)
+
+	if (!SSdbcore.Connect())
+		to_chat(
+			client,
+			type = MESSAGE_TYPE_ADMINLOG,
+			html = "<h1><b class='danger'>You could not be verified, and a DB connection couldn't be established. Please contact a higher admin to grant you permission.</b></h1>",
```

Comment:
I think it'd be better to explicitly say `contact someone with +permissions` as to a trial admin a gamemin is a higher admin and they wouldn't be able to help.